spring-kafka/src/test/java/org/springframework/kafka/core/KafkaTemplateTests.java

```diff
@@ -132,6 +172,46 @@ public void testTemplate() throws Exception {
 	}
 
 	@Test
+	public void testWithMessage() throws Exception {
+		Map<String, Object> senderProps = KafkaTestUtils.producerProps(embeddedKafka);
+		DefaultKafkaProducerFactory<Integer, String> pf = new DefaultKafkaProducerFactory<>(senderProps);
+		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(pf, true);
+
+		Message<String> message1 = MessageBuilder.withPayload("foo-message")
+				.setHeader(KafkaHeaders.TOPIC, INT_KEY_TOPIC)
+				.setHeader(KafkaHeaders.PARTITION_ID, 0)
+				.build();
+
+		template.send(message1);
+
+		ConsumerRecord<Integer, String> r1 = KafkaTestUtils.getSingleRecord(consumer, INT_KEY_TOPIC);
+		assertThat(r1).has(value("foo-message"));
+
+		Message<String> message2 = MessageBuilder.withPayload("foo-message-2")
+				.setHeader(KafkaHeaders.TOPIC, INT_KEY_TOPIC)
+				.setHeader(KafkaHeaders.PARTITION_ID, 0)
+				.setHeader(KafkaHeaders.TIMESTAMP, 1487694048615L)
+				.build();
+
+		template.send(message2);
+
+		ConsumerRecord<Integer, String> r2 = KafkaTestUtils.getSingleRecord(consumer, INT_KEY_TOPIC);
+		assertThat(r2).has(value("foo-message-2"));
+		assertThat(r2).has(timestamp(TimestampType.CREATE_TIME, 1487694048615L));
+
+		MessagingMessageConverter messageConverter = new MessagingMessageConverter();
+
+		Message<?> recordToMessage = messageConverter.toMessage(r2, null, String.class);
+
+		assertThat(recordToMessage.getHeaders().get(KafkaHeaders.TIMESTAMP_TYPE)).isEqualTo("CreateTime");
```

Comment:
I think we can use `TimestampType.CREATE_TIME.name()` for comparison,
But, yeah! That was clever do not populate Kafka-specific types into the `MessageHeaders` 😄